internal/db/db.go

```diff
@@ -42,42 +44,49 @@ func StringToDbType(dialect string) (DbType, error) {
 
 // DB is a wrapper around the ORM
 type DB struct {
-	mu      sync.Mutex
-	wrapped *dbw.DB
+	wrapped *atomic.Pointer[dbw.DB]
 }
 
-type closeDbFn func(context.Context) error
+type closeDbFn func(context.Context)
+
+var CloseSwappedDbDuration = 5 * time.Minute
 
 // Swap replaces *DB's underlying database object with the one from `newDB`.
 // It returns a function that calls Close() on the outgoing database object.
 // Note: Swap does not verify the incoming *DB object is correctly set-up.
 func (db *DB) Swap(ctx context.Context, newDB *DB) (closeDbFn, error) {
-	if newDB == nil || newDB.wrapped == nil {
+	const op = "db.(DB).Swap"
+	if newDB == nil || newDB.wrapped == nil || newDB.wrapped.Load() == nil {
 		return nil, fmt.Errorf("no new db object present")
 	}
-	if db.wrapped == nil {
+	if db == nil || db.wrapped == nil || db.wrapped.Load() == nil {
 		// TBD: We could be helpful here and set the new DB instead of err?
 		return nil, fmt.Errorf("no current db is present to swap, aborting")
 	}
 
 	// Grab the old db to allow for cleanup after swap.
-	oldDbw := *db.wrapped
-	closeOldDbFn := func(ctx context.Context) error { return oldDbw.Close(ctx) }
-
-	// Replace the current DB value with the new one.
-	// Note: It's important that the pointer doesn't change, rather the value
-	// the pointer points to. Changing the pointer itself does not update the
-	// references in the other objects (repo functions, etc).
-	db.mu.Lock()
-	*db.wrapped = *newDB.wrapped
-	db.mu.Unlock()
+	oldDbw := db.wrapped.Swap(newDB.wrapped.Load())
+	closeOldDbFn := func(ctx context.Context) {
+		go func() {
+			t := time.NewTimer(CloseSwappedDbDuration)
+			select {
+			case <-t.C:
+				if err := oldDbw.Close(ctx); err != nil {
+					event.WriteError(ctx, op, errors.Wrap(ctx, err, op))
+				}
```

Comment:
This is my only concern, because it means that for any swap operation, we'll have 2  connections (per controller) that will just sit there for whatever's set in `CloseSwappedDbDuration`.

We've already got a significant problem with database connections and this will only push our average connection count up over time on deployments that make use of this feature.

As it stands, if we change the database credentials and switch over, we'd have a minimum of 5 connections per controller (assuming we lock the DB) for whatever is in `CloseSwappedDbDuration`, after which will revert to the normal 3.

![image](https://user-images.githubusercontent.com/10965479/189337642-7786b42d-8069-4de8-88d8-6f80cf2df87c.png)

I don't particularly have a code solution for this, and I understand why we need to do this, I just thought it was something that we should be aware of. My only idea at the moment is that perhaps on deployments that make use of this feature we could tweak `CloseSwappedDbDuration` to minimise any impact this could have.

Reply:
We could try examining the [sql.DBStats](https://pkg.go.dev/database/sql#DBStats) to see if there is only idle connections in the pool and close instead of waiting the full time.